File: algorithms/lab08/python/03-bellman/03-bellman.py
# SPDX-License-Identifier: BSD-3-Clause


class Task:
    def __init__(self) -> None:
        # valoare mai mare decat orice distanta din graf
        self.INF = 1 << 30

        # n = numar de noduri, m = numar de muchii
        self.n = 0
        self.m = 0

        # adj[node] = lista de adiacenta a nodului node
        # perechea (neigh, w) semnifica arc de la node la neigh de cost w
        self.adj: list[list[tuple[int, int]]] = []

        # nodul sursa
        self.source = 0

    def solve(self) -> None:
        self.read_input()
        self.write_output(self.get_result())

    def read_input(self) -> None:
        with open("in", "r") as fin:
            tokens = fin.read().split()
            if not tokens:
                return

            self.n = int(tokens[0])
            self.m = int(tokens[1])
            self.source = int(tokens[2])

            self.adj = [[] for _ in range(self.n + 1)]

            idx = 3
            for _ in range(self.m):
                x = int(tokens[idx])
                y = int(tokens[idx + 1])
                w = int(tokens[idx + 2])
                self.adj[x].append((y, w))
                idx += 3

    def get_result(self) -> tuple[bool, list[int]]:
        # Construiesc un vector de muchii.
        edges: list[tuple[int, int, int]] = []

        for node in range(1, self.n + 1):
            for neigh, w in self.adj[node]:
                edges.append((node, neigh, w))

        return self.bellman(self.source, edges)

    def bellman(self, source: int, edges: list[tuple[int, int, int]]) -> tuple[bool, list[int]]:
        # Initializam vectorul de distante cu distante infinite.
        d = [self.INF] * (self.n + 1)
        p = [0] * (self.n + 1)

        # Setez sursa la distanta 0.
        d[source] = 0

        # Fac N - 1 relaxari.
        for _ in range(1, self.n):
            # Parcurg toate muchiile:
            for node, neigh, w in edges:
                # Se imbunatateste distanta?
                if d[node] + w < d[neigh]:
                    # Actualizam distanta si parintele.
                    d[neigh] = d[node] + w
                    p[neigh] = node

        # Verific daca mai poate fi updatata distanta.
        for node, neigh, w in edges:
            # Se imbunatateste distanta?
            if d[node] + w < d[neigh]:
                # Am gasit un ciclu de cost negativ.
                return (True, [])

        # Toate nodurile catre care distanta este inca INF nu pot fi atinse din
        # nodul source, deci le setam pe -1.
        for node in range(1, self.n + 1):
            if d[node] == self.INF:
                d[node] = -1

        return (False, d)

    def write_output(self, result: tuple[bool, list[int]]) -> None:
        has_cycle, d = result
        with open("out", "w") as fout:
            if has_cycle:
                fout.write("Ciclu negativ!\n")
            else:
                out_str = " ".join(str(d[node]) for node in range(1, self.n + 1))
                fout.write(out_str + " \n")


def main() -> None:
    task = Task()
    task.solve()


# [ATENTIE] NU modifica functia main!
if __name__ == "__main__":
    main()

File: algorithms/lab08/rust/03-bellman/03-bellman.rs
// SPDX-License-Identifier: BSD-3-Clause

use std::fs::File;
use std::io::{BufRead, BufReader, Write};

#[derive(Default)]
struct Task {
    // n = numar de noduri, m = numar de muchii
    n: i32,
    m: i32,

    // adj[node] = lista de adiacenta a nodului node
    // perechea (neigh, w) semnifica arc de la node la neigh de cost w
    adj: Vec<Vec<(i32, i32)>>,

    // nodul sursa
    source: i32,
}

struct Edge {
    node: i32,
    neigh: i32,
    w: i32,
}

impl Task {
    // numarul maxim de noduri
    const NMAX: usize = 50_005;

    // valoare mai mare decat orice distanta din graf
    const INF: i32 = 1 << 30;

    pub fn solve(&mut self) {
        self.read_input();
        let (has_cycle, d) = self.get_result();
        self.write_output(has_cycle, &d);
    }

    fn read_input(&mut self) {
        let fin = File::open("in").expect("in");
        let mut lines = BufReader::new(fin).lines().map(|l| l.unwrap());
        let first = lines.next().unwrap();
        let mut tokens = first.split_whitespace().map(|s| s.parse::<i32>().unwrap());
        self.n = tokens.next().unwrap();
        self.m = tokens.next().unwrap();
        self.source = tokens.next().unwrap();
        self.adj = vec![vec![]; Self::NMAX];
        for _ in 0..self.m {
            let line = lines.next().unwrap();
            let mut tok = line.split_whitespace().map(|s| s.parse::<i32>().unwrap());
            let x = tok.next().unwrap();
            let y = tok.next().unwrap();
            let w = tok.next().unwrap();
            self.adj[x as usize].push((y, w));
        }
    }

    fn get_result(&self) -> (bool, Vec<i32>) {
        // Construiesc un vector de muchii.
        let mut edges: Vec<Edge> = vec![];

        for node in 1..=self.n {
            for &(neigh, w) in &self.adj[node as usize] {
                edges.push(Edge { node, neigh, w });
            }
        }

        return self.bellman(self.source, &edges);
    }

    fn bellman(&self, source: i32, edges: &[Edge]) -> (bool, Vec<i32>) {
        // Initializam vectorul de distante cu distante infinite.
        let mut d = vec![Self::INF; (self.n + 1) as usize];

        // Setez sursa la distanta 0.
        d[source as usize] = 0;

        // Fac N - 1 relaxari.
        for _ in 1..self.n {
            // Parcurg toate muchiile:
            for edge in edges {
                // Se imbunatateste distanta?
                if d[edge.node as usize] + edge.w < d[edge.neigh as usize] {
                    // Actualizam distanta.
                    d[edge.neigh as usize] = d[edge.node as usize] + edge.w;
                }
            }
        }

        // Verific daca mai poate fi updatata distanta.
        for edge in edges {
            // Se imbunatateste distanta?
            if d[edge.node as usize] + edge.w < d[edge.neigh as usize] {
                // Am gasit un ciclu de cost negativ.
                return (true, vec![]);
            }
        }

        // Toate nodurile catre care distanta este inca INF nu pot fi atinse din
        // nodul source, deci le setam pe -1.
        for node in 1..=self.n {
            if d[node as usize] == Self::INF {
                d[node as usize] = -1;
            }
        }

        return (false, d);
    }

    fn write_output(&self, has_cycle: bool, d: &[i32]) {
        let mut fout = File::create("out").expect("out");
        if has_cycle {
            writeln!(fout, "Ciclu negativ!").unwrap();
        } else {
            for node in 1..=self.n {
                write!(fout, "{} ", d[node as usize]).unwrap();
            }
            writeln!(fout).unwrap();
        }
    }
}

// [ATENTIE] NU modifica functia main!
fn main() {
    let mut task = Box::new(Task::default());
    task.solve();
}
